### PR TITLE
Feature/spec validation

### PR DESCRIPTION
**Pull request by zonotope**
This (Draft) patch adds FlureeQL specs and uses them for validation. It also removes the `fluree.db.query.analytical` namespace because it wasn't used.

### CODE REVIEW

File: src/fluree/db/query/fql.cljc [hunk 1]
```diff
@@ -11,13 +15,198 @@
...
+                        :vector (s/cat :direction ::direction
+                                       :field     ::var)))
+
+(s/def ::orderBy (s/or :clause     ::ordering
+                       :collection (s/coll-of ::ordering)))
+
+(s/def ::groupBy (s/or :clause     ::var
+                       :collection (s/coll-of ::var)))
+
+(def first-key
```
Comment (dpetran) [inline src/fluree/db/query/fql.cljc:109]: Does this rely on small maps being array-maps? I'm trying to figure out if this is deterministic.

File: src/fluree/db/query/fql.cljc [hunk 2]
```diff
@@ -11,13 +15,198 @@
...
+  (and (list? x)
+       (-> x first symbol?)))
+
+(s/def ::function (s/or :string fn-string?, :list fn-list?))
+
+(s/def ::filter (s/coll-of ::function))
+
+(defn wildcard?
+  [x]
+  (#{"*" :* '*} x))
```
Comment (dpetran) [inline src/fluree/db/query/fql.cljc:55]: This isn't something we need to fix right now but I'm curious if there's a way to translate all queries/transactions into a single canonical representation so we don't have to match all three styles.

Comment (zonotope) [reply src/fluree/db/query/fql.cljc:55]: yep, that's in the plan. There's a `normalize` fn here that's called before validation. Eventually, when all the specs are complete, we'll do all the necessary `->kebab-cased-keyword` translation among other things as a first step before validation. The tension here is that I would also like to validate queries on the client before they're sent to the query peer as well, so we need to make sure we can handle all the cases that arise from user input in a clean and unsurprising way as well.

File: src/fluree/db/query/fql.cljc [hunk 3]
```diff
@@ -11,13 +15,198 @@
...
+
+(defmethod where-map-spec :default
+  [_]
+  (constantly false))
+
+(s/def ::where-map (s/and (s/map-of ::where-op map?, :count 1)
+                          (s/multi-spec where-map-spec first-key)))
+
+(s/def ::where-tuple (s/and sequential?
+                            (s/or :binding (s/coll-of (constantly true), :count 2)
```
[the remark this answers is not in the document]
Comment (zonotope) [reply src/fluree/db/query/fql.cljc:145]: why, yes they can!

File: src/fluree/db/query/fql.cljc (hunk 1 above)
Comment (zonotope) [reply src/fluree/db/query/fql.cljc:109]: I don't assume it's deterministic for maps with more than one entry, but I also do a [check to ensure that this map also has only one entry](https://github.com/fluree/db/pull/254/files#diff-305ee07335574d5476e0a2117e5895bc63b2086994b9e764dde30545c7c2c351R141) in the spec definition that uses this.

**Review by dpetran — APPROVED** [review]
:+1: I like seeing all that validation consolidated

**Comment by cap10morgan** [comment]
🙌🏻

**Comment by bplatz** [comment]
I do think, because this is such a critical performance part of our system, that we should do a criterium quick-bench on the parse phase as it stands and with this new process to understand the impact. Hopefully minimal as it is really nice to have this defined in a 'standard' way all together.

When we moved GraphQL parsing off of instaparse and onto just native code it was remarkably faster. Spec might be more efficient than instaparse, but I think of it as a similar type of abstraction... and its design goal was more around development and testing, not end-user input validation, so performance might not even be a design consideration.

**Comment by zonotope** [comment]
> I do think, because this is such a critical performance part of our system, that we should do a criterium quick-bench on the parse phase as it stands and with this new process to understand the impact. Hopefully minimal as it is really nice to have this defined in a 'standard' way all together.
> 
> When we moved GraphQL parsing off of instaparse and onto just native code it was remarkably faster. Spec might be more efficient than instaparse, but I think of it as a similar type of abstraction... and its design goal was more around development and testing, not end-user input validation, so performance might not even be a design consideration.

Yep, I think this definitely could be a concern. I thought it would be ok because I thought that the amount of time to actually fetch and process the flakes for the query would dwarf whatever time was spent in the parser, but I didn't consider the impact of cached index nodes.

I was planning on keeping this pr in draft status while I worked on implementing the rest of the parser so I could identify other nuances of the format from the code. I will definitely set up a benchmark and post the results here after that for further discussion.

**Comment by zonotope** [comment]
Closing because this work has been integrated as part of #304

**Pull request closed without merging**